packages/peregrine/lib/talons/CartPage/GiftOptions/useGiftOptions.js

```diff
@@ -0,0 +1,206 @@
+import { useState, useCallback, useEffect, useMemo } from 'react';
+import throttle from 'lodash.throttle';
+import gql from 'graphql-tag';
+import { useMutation, useQuery } from '@apollo/react-hooks';
+
+import { useCartContext } from '@magento/peregrine/lib/context/cart';
+import { useRestApi } from '@magento/peregrine/lib/hooks/useRestApi';
+import { useUserContext } from '@magento/peregrine/lib/context/user';
+
+/**
+ * Local query. GQL support is not available as of today.
+ *
+ * Once available, we can change the query to match the schema.
+ */
+const GET_GIFT_OPTIONS_QUERY = gql`
+    query getGiftOptions {
+        gift_options @client {
+            include_gift_receipt
+            include_printed_card
+            gift_message
+        }
+    }
+`;
+
+/**
+ * Local mutation. GQL support is not available as of today.
+ *
+ * Once available, we can change the mutation to match the schema.
+ */
+const SET_GIFT_OPTIONS_QUERY = gql`
+    mutation setGiftOptions(
+        $cart_id: String
+        $include_gift_receipt: Boolean
+        $include_printed_card: Boolean
+        $gift_message: String
+    ) {
+        set_gift_options(
+            cart_id: $cart_id
+            include_gift_receipt: $include_gift_receipt
+            include_printed_card: $include_printed_card
+            gift_message: $gift_message
+        ) @client
+    }
```

Comment:
Your custom mutation should apply the new options to the cart cache and then return that entire object. See places where we use `...CartPageFragment`.

Reply:
Cart cache is temp, only till the GQL team comes up with gift options coverage. I don't think we should be using it.

Additionally, waiting for a whole mutation/REST call to check or uncheck a box, is very snappy and does not provide a good UI experience.

Reply:
I can see the confusion here. If gift options are a single operation that are guaranteed not to modify the cart prices or other values then this is fine. I'm not sure what the docs say about it with REST. Do you just make a single call (apply gift options) and continue with checkout? Or is there a second step like "apply gift options **and then** fetch the new total"?

Reply:
We should be calling the gift options rest call when the user clicks on checkout. Till then, it will be stored in the apollo cache.